Create new episode for pull push (#88)
ImperialCollegeLondon/intermediate_grad_school_git_course: _episodes/05-pulling_pushing.md

---
title: "Pulling and Pushing"
teaching: 10
exercises: 5
questions:
- "How do I keep my local branches in sync with the remote branches?"
- "What is the difference between forking and branching?"
objectives:
- Push changes to a remote repository.
- Pull changes from a remote repository.
keypoints:
- With `git push` you can push any committed changes in your current branch to its upstream branch.
- If the current branch has no upstream yet, you can configure one by doing `git push -u origin BRANCH_NAME`.
- Using `git pull` will bring changes in the upstream branch to the local branch.
---

> ## Multiple branches in remotes

Reviewer: ```suggestion
## Multiple remote branches
```

> The same way you might have different branches in your local repository, you could
> manage different branches in your remote - the same branches or different ones.

Reviewer: How about:

```suggestion
Just as you can have multiple local branches, you can also have multiple remote branches. These may or may not be upstreams for your local branches.
```

(...or something. I'm not sure.)

```sh
git switch main
```
{: .commands}
```
Switched to branch 'main'
Your branch is ahead of 'origin/main' by 6 commits.
  (use "git push" to publish your local commits)
```
{: .output}

Notice that git is suggesting you to use `git push` to publish your local commits. Let's do that:

```sh
git push
```
{: .commands}
```
Enumerating objects: 21, done.
Counting objects: 100% (21/21), done.
Delta compression using up to 8 threads
Compressing objects: 100% (18/18), done.
Writing objects: 100% (18/18), 1.83 KiB | 938.00 KiB/s, done.
Total 18 (delta 4), reused 0 (delta 0), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (4/4), completed with 1 local object.
To https://github.com/username/recipe.git
   57d4505..d10e1e9  main -> main
```
{: .output}

> ## Now you try
>
> You should now be on the `main` branch. Try switching to the `spicy`
> branch and pushing changes to it.
>
> 1. Check the current branch using `git branch`.
> 1. If the current branch is `main`, then switch to `spicy` using `git switch spicy`.
> 1. Push changes to the `spicy` branch using `git push`.
>
> > ## Solution
> >
> > ```
> > git branch
> > git switch spicy
> > git push
> > ```
> > {: .commands}

> > > This should give rise to an error that the current branch `spicy` has no upstream branch:

Reviewer: ```suggestion
> > This should give an error that the current branch `spicy` has no upstream branch:
```

> > ```
> > fatal: The current branch spicy has no upstream branch.
> > To push the current branch and set the remote as upstream, use
> >
> >     git push --set-upstream origin spicy
> >
> > To have this happen automatically for branches without a tracking
> > upstream, see 'push.autoSetupRemote' in 'git help config'.
> > ```
> > Push again by setting the upstream:
> > ```

> > > git push --set-upstream origin spicy

Reviewer: I think earlier in the text we tell students to use the `-u` flag instead. Maybe it would be worth mentioning the `--set-upstream` flag when they first encounter it and telling them that `-u` is a shortcut for it?

Reviewer 2: Yes, I agree. Put this one earlier and mention `-u` can be used instead

> > Writing objects: 100% (3/3), 347 bytes | 347.00 KiB/s, done.
> > Total 3 (delta 1), reused 0 (delta 0), pack-reused 0 (from 0)
> > remote: Resolving deltas: 100% (1/1), completed with 1 local object.
> > remote:
> > remote: Create a pull request for 'spicy' on GitHub by visiting:
> > remote:      https://github.com/username/recipe/pull/new/spicy
> > remote:
> > To https://github.com/username/recipe.git
> >  * [new branch]      spicy -> spicy
> > branch 'spicy' set up to track 'origin/spicy'.
> > ```
> > {: .output}
> {: .solution}
{: .challenge}

## Pulling

* Opposite to `push`, `pull` brings changes in the upstream branch to the local branch.
* You can check if there are any changes to synchronise in the upstream branch by
  running `git fetch`, which only checks if there are changes, and then `git status` to
  see how your local and remote branch compare in terms of commit history.
* It's best to make sure your repository is in a clean state with no staged or unstaged
  changes.
* If the local and upstream branches have diverged - have different commit history - the
  command will attempt to merge both. If there are conflicts, you will need deal with

>   them in the same way described above.

Reviewer: Should we also mention this in the `keypoints` at the start?

> * You can get a new branch existing only in `origin` directly with `git switch
>   BRANCH_NAME` without the need of creating the branch locally and then pulling the
>   remote.

Reviewer: Maaaaaaybe something like this:

```suggestion
* You can get a new branch that exists only in `origin` directly with `git switch
  BRANCH_NAME` which will automatically create a local branch with the same name
```

I'm not sure my suggestion is necessarily clearer though :shrug:

Reviewer 2: I think Alex wording is a little clearer just because it is shorter

> > ```
> > Push again by setting the upstream:
> > ```

> > > git push --u origin spicy

Reviewer 2: This should be a single dash, i.e.

`git push -u origin spicy`
